spec/compiler/semantic/abstract_def_spec.cr

```diff
@@ -584,6 +584,358 @@ describe "Semantic: abstract def" do
+      ))
+  end
+
+  it "matches free var with itself, inside generics" do
+    semantic(%(
+      module Val(T); end
+
+      module Foo
+        abstract def foo(x : Val(T)) forall T
+      end
+
+      class Bar
+        include Foo
+
+        def foo(x : Val(T)) forall T
+        end
+      end
+      ))
+  end
+
+  it "matches free var with renamed free var" do
+    semantic(%(
+      module Foo
+        abstract def foo(x : T) forall T
+      end
+
+      class Bar
+        include Foo
+
+        def foo(x : U) forall U
+        end
+      end
+      ))
+  end
+
+  it "matches free var with renamed free var, inside generics" do
+    semantic(%(
+      module Val(T); end
+
+      module Foo
+        abstract def foo(x : Val(T)) forall T
+      end
+
+      class Bar
+        include Foo
+
+        def foo(x : Val(U)) forall U
+        end
+      end
+      ))
+  end
+
+  it "errors if underscore is not matched" do
+    assert_error %(
+      module Foo
+        abstract def foo(x : _)
+      end
+
+      class Bar
+        include Foo
+
+        def foo(x : Int32)
+        end
+      end
+      ), "abstract `def Foo#foo(x : _)` must be implemented by Bar"
+  end
+
+  it "errors if free var is not matched" do
+    assert_error %(
+      module Foo
+        abstract def foo(x : T) forall T
+      end
+
+      class Bar
+        include Foo
+
+        def foo(x : Int32)
+        end
+      end
+      ), "abstract `def Foo#foo(x : T)` must be implemented by Bar"
+  end
+
+  it "errors if generic free var is not matched" do
+    assert_error %(
+      module Val(T); end
+
+      module Foo
+        abstract def foo(x : Val(T)) forall T
+      end
+
+      class Bar
+        include Foo
+
+        def foo(x : Val(Int32))
+        end
+      end
+      ), "abstract `def Foo#foo(x : Val(T))` must be implemented by Bar"
+  end
+
+  it "matches more constrained free var with a less constrained one" do
+    semantic(%(
+      module Val(T); end
+
+      module Foo
+        abstract def foo(x : Val(T), y : Val(Val(_))) forall T
+      end
+
+      class Bar
+        include Foo
+
+        def foo(x : T, y : Val(_)) forall T
+        end
+      end
+      ))
+  end
+
+  it "errors if free var is more constrained than abstract def" do
+    assert_error %(
+      module Val(T); end
+
+      module Foo
+        abstract def foo(x : Val(T)) forall T
+      end
+
+      class Bar
+        include Foo
+
+        def foo(x : Val(Val(T))) forall T
+        end
+      end
+      ), "abstract `def Foo#foo(x : Val(T))` must be implemented by Bar"
+  end
+
+  it "errors if free var is bound in including type" do
+    assert_error %(
+      module Foo
+        abstract def foo(x : T) forall T
+      end
+
+      class Bar(T)
+        include Foo
+
+        def foo(x : T)
+        end
+      end
+      ), "abstract `def Foo#foo(x : T)` must be implemented by Bar(T)"
+  end
+
+  it "errors if free vars in implementation are less constrained than abstract def (1)" do
```

Comment:
Isn't it the other way around? Free vars in implementation are more constrained because they are equal instead of independend. Ditto below.
Specs themselves are good, though 👍